src/mcp-server-factory/McpServerFactory.ts

```diff
@@ -0,0 +1,243 @@
+ */
+export type ServerMode = 'DYNAMIC_TOOL_DISCOVERY' | 'STATIC_TOOL_SETS' | 'ALL_TOOLS';
+
+/**
+ * Session configuration type matching the SDK's CreateServerArg config
+ */
+export interface SessionConfig {
+  FMP_ACCESS_TOKEN?: string;
+  FMP_TOOL_SETS?: string;
+  DYNAMIC_TOOL_DISCOVERY?: string;
+}
+
+/**
+ * Configuration options for server creation
+ * Compatible with SDK's CreateServerArg<T> where T = SessionConfig
+ */
+export interface McpServerOptions {
+  sessionId: string;
+  config?: SessionConfig;
+  serverAccessToken?: string;
+}
+
+/**
+ * Result of server creation
+ */
+export interface McpServerCreationResult {
+  mcpServer: McpServer;
+  toolManager?: DynamicToolsetManager; // Only present in DYNAMIC_TOOL_DISCOVERY mode
+  mode: ServerMode;
+}
+
+/**
+ * Factory class for creating and configuring MCP servers
+ * Handles server creation logic and tool registration based on mode
+ */
+export class McpServerFactory {
+  private readonly version: string;
+
+  constructor() {
+    this.version = getServerVersion();  
+  }
+
+  /**
+   * Creates a new MCP server with the appropriate configuration and tools
+   * Compatible with SDK's CreateServerFn<SessionConfig> signature
+   * @param options - Server creation options
+   * @returns Server creation result with server, manager, and mode
+   */
+  public createServer(options: McpServerOptions): McpServerCreationResult {
+    const { sessionId, config, serverAccessToken } = options;
+
+    // Resolve access token and mode first
+    const accessToken = this._resolveAccessToken(serverAccessToken, config);
+    const mode = this._resolveSessionMode(config);
+    
+    console.log(`[McpServerFactory] Creating server for session ${sessionId} in ${mode} mode`);
+
+    // Create the base MCP server with appropriate capabilities
+    const isDynamicMode = mode === 'DYNAMIC_TOOL_DISCOVERY';
+    const mcpServer = this._createMcpServerInstance(isDynamicMode);
+
+    // Register tools based on mode
+    const toolManager = this._registerToolsForMode(mcpServer, mode, config, accessToken);
+
+    return {
+      mcpServer,
+      toolManager,
+      mode
+    };
+  }
+
+  /**
+   * Creates a new, isolated McpServer instance with proper configuration schema
+   * @param isDynamicMode - Whether this server instance supports dynamic tool changes
+   * @returns Configured McpServer instance
+   * 
+   * Note: There's a known issue in @modelcontextprotocol/sdk v1.0.0 where the main
+   * capabilities.tools.listChanged field may not reflect the server configuration correctly.
+   * The serverInfo.capabilities.tools.listChanged field does reflect the correct value.
+   */
+  private _createMcpServerInstance(isDynamicMode: boolean = false): McpServer {
+    return new McpServer({
+      name: "Financial Modeling Prep MCP (Stateful)",
+      version: this.version,
+      capabilities: {
+        tools: { 
+          listChanged: isDynamicMode // Only enable dynamic tool changes in dynamic mode
```

Comment:
Do you have info on which MCP clients already supports this? (I'm assuming it's not just "magically" supported)

Reply:
its supported by  mcp client sdk 1.0.0 and above